crates/sui-verifier/src/char_type_verifier.rs

```diff
@@ -0,0 +1,227 @@
+    if ModuleId::new(SUI_FRAMEWORK_ADDRESS, ident_str!("sui").to_owned()) == module.self_id() {
+        return Ok(());
+    }
+
+    let view = BinaryIndexedView::Module(module);
+    let mod_handle = view.module_handle_at(module.self_module_handle_idx);
+    let mod_name = view.identifier_at(mod_handle.name).as_str();
+    let struct_defs = &module.struct_defs;
+    let mut char_type_candidate = None;
+    for def in struct_defs {
+        let struct_handle = module.struct_handle_at(def.struct_handle);
+        let struct_name = view.identifier_at(struct_handle.name).as_str();
+        if struct_name.to_ascii_uppercase() == struct_name
+            && mod_name.to_ascii_uppercase() == struct_name
+        {
+            verify_char_type(module, struct_name, struct_handle, def)
+                .map_err(verification_failure)?;
+            // if we reached this point, it means we have a legitimate characteristic type candidate
+            // and we have to make sure that both the init function's signature reflects this and
+            // that this type is not instantiated in any function of the module
+            char_type_candidate = Some((struct_name, struct_handle, def));
+            break; // no reason to look any further
+        }
+    }
+    for fn_def in &module.function_defs {
+        let fn_handle = module.function_handle_at(fn_def.function);
+        let fn_name = module.identifier_at(fn_handle.name);
+        if fn_name == INIT_FN_NAME {
+            if let Some((struct_name, struct_handle, _)) = char_type_candidate {
+                // only verify if init function conforms to characteristic types requirements if we
+                // have a characteristic type candidate
+                verify_init_function_char_type(module, fn_handle, struct_name, struct_handle)
+                    .map_err(verification_failure)?;
+            } else {
+                // if there is no characteristic type candidate than the init function should have
+                // only one parameter of TxContext type
+                verify_init_function_single_param(module, fn_handle)
+                    .map_err(verification_failure)?;
+            }
+        }
+        if let Some((struct_name, _, def)) = char_type_candidate {
+            // only verify lack of characteristic types instantiations if we have a
+            // characteristic type candidate
+            verify_no_instantiations(module, fn_def, struct_name, def)
+                .map_err(verification_failure)?;
+        }
+    }
+
+    Ok(())
+}
+
+// Verifies all required properties of a characteristic type candidate (that is a type whose name is
+// the same as the name of a
+fn verify_char_type(
+    module: &CompiledModule,
+    struct_name: &str,
+    struct_handle: &StructHandle,
+    struct_def: &StructDefinition,
+) -> Result<(), String> {
+    // must have only one ability: drop
+    let drop_set = AbilitySet::from_u8(Ability::Drop as u8).unwrap();
```

Comment:
`AbilitySet::Empty | Ability::Drop` should work

Reply:
Thank you! I felt that there should be a better way...